Can't resolve '@ngtools/webpack' in yarn monorepo

<!--🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅

Oh hi there! 😄

To expedite issue processing please search open and closed issues before submitting a new one.
Existing issues often contain information about workarounds, resolution, or progress updates.

🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅-->


# 🐞 Bug report

### Command (mark with an `x`)
<!-- Can you pin-point the command or commands that are effected by this bug? -->
<!-- ✍️edit: -->
```
- [ ] new
- [x] build
- [x] serve
- [x] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Is this a regression?

<!-- Did this behavior use to work in the previous version? -->
No. (Perhaps in a really old version before migrating to the monorepo structure, see more in description)


### Description

During the build process (on `build`, `serve` or `test`) an error occurs (see below) and the build fails. The `@ngtools/webpack` package can not be resolved. This happens due to a yarn monorepo structure, in which `@ngtools/webpack` will not be placed in the root `./node_modules` folder due conflicts with other dependencies, so it gets placed under `./node_modules/@angular-devkit/build-angular/node_modules`. And then it can not be found during build.

I tracked down the issue and found the source. You can see the problem also in issue #10220. It was fixed by @hansl in the old single repo: https://github.com/angular/devkit/pull/647 but didn't get copied to the new monorepo? --> https://github.com/angular/angular-cli/blob/master/packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/typescript.ts#L23

**Can we do this change once again?**

If I change those lines locally the build is successful, but I do not know if we can simply change this without other problems?

Note: If you have a monorepo with only one angular application this problem should not happen, but in my case the monorepo has also some angularjs applications and multiple webpack configurations. So multiple webpack, loader and angular dependency versions are around, not all aligned.


## 🔬 Minimal Reproduction

Reproduction repo: https://github.com/fabianbru/angular-cli-bug-reproduction
This repo is a striped down repo, so there are many dependencies without real code.

Step-by-Step guide:
- Node 10 (Node version should not matter)
- yarn 1.13.0
- Checkout
- run `yarn` in the root to install all dependencies
- run `yarn dev` in the root
--> ERROR will occur
- change lines 7-9 locally in `node_modules/@angular-devkit/build-angular/src/angular-cli-files/models/webpack-configs/typescript.js` to:
```
const webpackLoader = require.resolve('@ngtools/webpack')
```
---> build should be successful

## 🔥 Exception or Error
<pre><code>
<!-- If the issue is accompanied by an exception or an error, please share it below: -->
ERROR in multi <workspace-path>/angular-cli-bug-reproduction/node_modules/webpack-dev-server/client?http://0.0.0.0:0 webpack/hot/dev-server ./src/main.ts
Module not found: Error: Can't resolve '@ngtools/webpack' in '<workspace-path>/angular-cli-bug-reproduction/packages/angular'
ERROR in multi ./src/polyfills.ts
Module not found: Error: Can't resolve '@ngtools/webpack' in '<workspace-path>/angular-cli-bug-reproduction/packages/angular'

<workspace-path> is the path to my workspace, like `/Users/xyz/workspace`
</code></pre>


## 🌍 Your Environment
<pre><code>
<!-- run `ng version` and paste output below -->
Angular CLI: 7.2.3
Node: 10.15.0
OS: darwin x64
Angular: 7.2.2
... animations, cdk, common, compiler, compiler-cli, core, forms
... http, language-service, material, platform-browser
... platform-browser-dynamic, platform-server, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.12.3
@angular-devkit/build-angular     0.12.3
@angular-devkit/build-optimizer   0.12.3
@angular-devkit/core              7.2.3
@angular-devkit/schematics        7.2.3
@angular/cli                      7.2.3
@angular/flex-layout              7.0.0-beta.23
@schematics/angular               7.2.3
@schematics/update                0.12.3
rxjs                              6.3.3
typescript                        3.2.4
webpack                           4.29.0
</code></pre>

**Anything else relevant?**

--- a/packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/typescript.ts
+++ b/packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/typescript.ts
@@ -13,16 +13,11 @@ import * as path from 'path';
 import {
   AngularCompilerPlugin,
   AngularCompilerPluginOptions,
+  NgToolsLoader,
   PLATFORM
 } from '@ngtools/webpack';
 import { buildOptimizerLoader } from './common';
 import { WebpackConfigOptions } from '../build-options';
-
-
-const g: any = typeof global !== 'undefined' ? global : {};
-const webpackLoader: string = g['_DevKitIsLocal']
-  ? require.resolve('@ngtools/webpack')
-  : '@ngtools/webpack';
 
 
 function _createAotPlugin(
@@ -92,7 +87,7 @@ export function getNonAotConfig(wco: WebpackConfigOptions, host: virtualFs.Host<
   const { tsConfigPath } = wco;
 
   return {
-    module: { rules: [{ test: /\.tsx?$/, loader: webpackLoader }] },
+    module: { rules: [{ test: /\.tsx?$/, loader: NgToolsLoader }] },
     plugins: [_createAotPlugin(wco, { tsConfigPath, skipCodeGeneration: true }, host)]
   };
 }
@@ -104,7 +99,7 @@ export function getAotConfig(
 ) {
   const { tsConfigPath, buildOptions } = wco;
 
-  const loaders: any[] = [webpackLoader];
+  const loaders: any[] = [NgToolsLoader];
   if (buildOptions.buildOptimizer) {
     loaders.unshift({
       loader: buildOptimizerLoader,
@@ -124,7 +119,7 @@ export function getNonAotTestConfig(wco: WebpackConfigOptions, host: virtualFs.H
   const { tsConfigPath } = wco;
 
   return {
-    module: { rules: [{ test: /\.tsx?$/, loader: webpackLoader }] },
+    module: { rules: [{ test: /\.tsx?$/, loader: NgToolsLoader }] },
     plugins: [_createAotPlugin(wco, { tsConfigPath, skipCodeGeneration: true }, host, false)]
   };
 }

--- a/packages/ngtools/webpack/src/index.ts
+++ b/packages/ngtools/webpack/src/index.ts
@@ -8,3 +8,5 @@
 
 export * from './angular_compiler_plugin';
 export { ngcLoader as default } from './loader';
+
+export const NgToolsLoader = __filename;

